### PR TITLE
Transformer moe [IPR - Do not review]

### PR DESCRIPTION
**Pull request by etragas-fathom**

### CODE REVIEW

File: tensor2tensor/models/research/transformer_moe.py [hunk 1]
```diff
@@ -122,8 +126,8 @@ def print_layer(name, layers):
     # ========= Construct the transformer encoder and decoder =========
 
     encoder_outputs = []
-
-    x = encoder_input
+    if encoder_input:
```
Comment (cbockman) [inline tensor2tensor/models/research/transformer_moe.py]: why is this necessary?

File: tensor2tensor/models/research/transformer_moe.py [hunk 2]
```diff
@@ -65,21 +65,26 @@ def body_sharded(self, sharded_features):
...
     inputs = sharded_features["inputs"]
     target_space = sharded_features["target_space_id"]
-    (
-        encoder_input,
-        encoder_self_attention_bias,
-        encoder_decoder_attention_bias,
-    ) = dp(self._prepare_encoder, inputs, target_space)
+
+    encoder_input = None
+    if encoder_layers:
```
Comment (cbockman) [inline tensor2tensor/models/research/transformer_moe.py]: what would happen if we didn't have this 'if' block?

File: tensor2tensor/models/research/transformer_moe.py [hunk 3]
```diff
@@ -65,21 +65,26 @@ def body_sharded(self, sharded_features):
...
+          encoder_input,
+          encoder_self_attention_bias,
+          encoder_decoder_attention_bias,
+      ) = dp(self._prepare_encoder, inputs, target_space)
 
     # Process output
     targets = sharded_features["targets"]
-    decoder_input, decoder_self_attention_bias = dp(
+    decoder_input = None
+    if decoder_layers:
```
Comment (cbockman) [inline tensor2tensor/models/research/transformer_moe.py]: what happens if this isn't in an 'if' block?

File: tensor2tensor/models/research/transformer_moe.py (hunk 2 above)
Comment (etragas-fathom) [reply tensor2tensor/models/research/transformer_moe.py]: Fair question. I think nothing would break. I added this to have similar behaviour for cases where only decoder or encoder are enabled. Though outside of symmetry I don't think this flag is necessary so I should remove it.

File: tensor2tensor/models/research/transformer_moe.py (hunk 3 above)
Comment (etragas-fathom) [reply tensor2tensor/models/research/transformer_moe.py]: We get an error in prepare_decoder 
```
  File "/usr/src/t2t/tensor2tensor/utils/t2t_model.py", line 231, in call
    sharded_logits, losses = self.model_fn_sharded(sharded_features)
  File "/usr/src/t2t/tensor2tensor/utils/t2t_model.py", line 273, in model_fn_sharded
    self._to_single_features_dict(transformed_features))
  File "/usr/src/t2t/tensor2tensor/models/research/transformer_moe.py", line 81, in body_sharded
    self._prepare_decoder, targets
  File "/usr/src/t2t/tensor2tensor/utils/expert_utils.py", line 231, in __call__
    outputs.append(fns[i](*my_args[i], **my_kwargs[i]))
  File "/usr/src/t2t/tensor2tensor/utils/expert_utils.py", line 60, in decorated
    return f(*args, **kwargs)
  File "/usr/src/t2t/tensor2tensor/models/research/transformer_moe.py", line 205, in _prepare_decoder
    targets, self._hparams, features=None,
  File "/usr/src/t2t/tensor2tensor/models/transformer.py", line 1209, in transformer_prepare_decoder
    decoder_input = common_attention.add_timing_signal_1d(decoder_input)
  File "/usr/src/t2t/tensor2tensor/utils/expert_utils.py", line 60, in decorated
    return f(*args, **kwargs)
  File "/usr/src/t2t/tensor2tensor/layers/common_attention.py", line 447, in add_timing_signal_1d
    return x + signal
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/math_ops.py", line 812, in binary_op_wrapper
    return func(x, y, name=name)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/gen_math_ops.py", line 365, in add
    "Add", x=x, y=y, name=name)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/op_def_library.py", line 547, in _apply_op_helper
    inferred_from[input_arg.type_attr]))
TypeError: Input 'y' of 'Add' Op has type float32 that does not match type int32 of argument 'x'.
```

File: tensor2tensor/models/research/transformer_moe.py (hunk 1 above)
Comment (etragas-fathom) [reply tensor2tensor/models/research/transformer_moe.py]: This is unnecessary, it was added for symmetry between behaviour of encoder_input and decoder_input.

File: tensor2tensor/models/research/transformer_moe.py [hunk 4]
```diff
@@ -101,12 +103,10 @@ def decorated(x, *args, **kwargs):
 
     # ========= Compute the transformer architecture =========
 
-    encoder_layers, decoder_layers = self._extract_layer_types()
 
     layers = common_attention.get_standardized_layers(
         hparams=hparams,
         dp=dp,
-        ps_devices=self._ps_devices,
```
Comment (cbockman) [inline tensor2tensor/models/research/transformer_moe.py]: why removed?

File: tensor2tensor/models/research/transformer_moe.py [hunk 5]
```diff
@@ -144,7 +144,8 @@ def print_layer(name, layers):
       if encoder_outputs:
         encoder_outputs[-1] = dp_preprocess(x)
 
-    x = decoder_input
+    if decoder_input:
+      x = decoder_input
     with tf.variable_scope("decoder"):
```
Comment (cbockman) [inline tensor2tensor/models/research/transformer_moe.py]: OK, this is getting more radical now, but now that we've straightened things out a bit:

* I think moving L149 - L177 into the indent makes sense
* can L82-84 just move under here, as well?
* wouldn't the 'if' being 'if decoder_layers' make more sense?
* if we do all of this, we end up with a big indent, which is unfortunate...probably this block as a func makes more sense...but prob not necessary no

Comment (cbockman) [reply tensor2tensor/models/research/transformer_moe.py]: (not sure how messy to pull this out into a func)

Comment (cbockman) [reply tensor2tensor/models/research/transformer_moe.py]: Note: also, moving L82-84 under would allow us to put back extract_layer_types where it was before

File: tensor2tensor/models/research/transformer_moe.py (hunk 4 above)
Comment (etragas-fathom) [reply tensor2tensor/models/research/transformer_moe.py]: Our defn of that fn doesn't accept ps_devices as an arg. https://github.com/medicode/tensor2tensor/blob/master/tensor2tensor/layers/common_attention.py#L46

I presume this code was broken in master when we last forked?

File: tensor2tensor/models/research/transformer_moe.py (hunk 5 above)
Comment (etragas-fathom) [reply tensor2tensor/models/research/transformer_moe.py]: Hmm, why bother indenting this code though? The for loop will try to enumerate decoder_layers which is an empty list if we've only got encoder enabled.
